File: c++/optional_test.cpp
#include <iostream>
#include <string>
#include <optional>

/**
 * The "test" cases use the same names as the test methods in the Java code, 
 * only styled to match C++ conventions.
 * 
 * The numbers in parenthesis, e.g. (1) or (2) match the Java code samples
 * in the respective test cases. Missing numbers mean missing features.
 * 
 * This pattern is used where applicable.
 */

/**
 * The sample class that will be used in all snippets.
 */
struct car 
{
    car(std::string man, std::string mod) 
        : manufacturer(man)
        , model(mod)
    {}

    bool operator==(const car& other) const 
    {
        return manufacturer == other.manufacturer && model == other.model;
    }

    friend std::ostream& operator<<(std::ostream& os, const car& car);

    std::string manufacturer;
    std::string model;
};

/**
 * The other sample class that will be used in a few snippets.
 */
struct truck
{
    truck(std::string man, std::string mod, std::string size) 
        : manufacturer(man)
        , model(mod)
        , size(size)
    {}

    bool operator==(const truck& other) const 
    {
        return manufacturer == other.manufacturer 
                && model == other.model 
                && size == other.size;
    }

    friend std::ostream& operator<<(std::ostream& os, const truck& car);

    std::string manufacturer;
    std::string model;
    std::string size;
};


std::ostream& operator<<(std::ostream& os, const car& car)
{
    os << car.manufacturer << " " << car.model;
    return os;
}


std::ostream& operator<<(std::ostream& os, const truck& truck)
{
    os << truck.size << " " << truck.manufacturer << " " << truck.model;
    return os;
}

static const auto loud_car = car{"Hyundai", "i30 N"};
static const auto reasonable_car = car{"Volkswagen", "Golf"};
static const auto huge_car = car{"Ford", "Raptor"};

/**
 * Demonstrate how std::nullopt can be used as a return value because std::optional
 * has a constructor that accepts this. Otherwise std::nullopt is not really useful
 * on its own.
 */
std::optional<car> make_empty_car() 
{
    return std::nullopt;
}

std::optional<car> produce_truck() 
{
    return std::make_optional(huge_car);
}

// Explicit types used for clarity.
void optional_of_null()
{
    std::cout << "optional_of_null" << std::endl
              << "----------------" << std::endl;
    
    // Does not work, requires object.
    // auto null_opt = std::make_optional(); 
    // Doesn't help us much by its own.
    std::nullopt_t null_opt = std::nullopt;
    // This on the other hand does.
    std::optional<car> null_opt_const = std::nullopt;
    // Or as a return from a method.
    std::optional<car> null_opt_method = make_empty_car();
    // Empty optional by creating a new std::optional manually.
    std::optional<car> null_opt_obj1 = std::optional<car>();
    // Empty optional by passing nullopt_t. Very long version of (2).
    std::optional<car> null_opt_obj2 = std::optional<car>(null_opt);

    // All work literally the same.
    if (!null_opt_const.has_value()) {
        std::cout << "null_opt_const has no value" << std::endl;
    }
    
    if (!null_opt_method.has_value()) {
        std::cout << "null_opt_method has no value" << std::endl;
    }
    
    if (!null_opt_obj1.has_value()) {
        std::cout << "null_opt_obj1 has no value" << std::endl;
    }

    if (!null_opt_obj2.has_value()) {
        std::cout << "null_opt_obj2 has no value" << std::endl;
    }
}

// auto types used for laziness. And because "auto" means "car" in German.
void optional_of_null_usage()
{
    std::cout << "optional_of_null_usage" << std::endl
              << "----------------------" << std::endl;

    // Given
    auto null_opt = make_empty_car();

    // When

    // (2) Get the actual value of the optional or another value if null.
    auto null_value_or = null_opt.value_or(reasonable_car);
    // (5) value() on empty optional throws.
    try {
        auto null_get = null_opt.value();
    } catch(const std::bad_optional_access& e) {
        std::cout << "value() throws if optional is empty" << std::endl;
    }
    

    // Then

    // (2)
    if (null_value_or == reasonable_car) {
        std::cout << "value_or() gets default object if optional is empty" << std::endl;
    }
}

void optional_of_value_usage()
{
    std::cout << "optional_of_value_usage" << std::endl
              << "-----------------------" << std::endl;

    // Given
    auto hyunday_opt = std::make_optional(loud_car);

    // When

    // (2)
    auto hyundai_value_or = hyunday_opt.value_or(reasonable_car);

    // Then

    // (2)
    if (hyundai_value_or == loud_car) {
        std::cout << "value_or() gets object if optional is not empty" << std::endl;
    }
}

void optional_of_value_other_usage()
{
    std::cout << "optional_of_value_other_usage" << std::endl
              << "-----------------------------" << std::endl;

    std::cout << "There's nothing std::optional has to offer here" << std::endl;
}

void bouncer_patterns()
{
    std::cout << "bouncer_patterns" << std::endl
              << "----------------" << std::endl;

    // Given
    auto hyunday_opt = std::make_optional(loud_car);

    // When

    // (1) Use has_value() condition. Most obvious and flexible approach.
    if (!hyunday_opt.has_value()) {
        // throw ...
        std::cout << "ERROR: If you see that then something is wrong" << std::endl;
    }
    auto value = hyunday_opt->manufacturer;
    std::cout << "Hyunday manufacturer value is " << value << " (surprise)" << std::endl;
}

void fake_repository_return_value()
{
    std::cout << "fake_repository_return_value" << std::endl
              << "----------------------------" << std::endl;

    std::cout << "Didn't bother to write helper methods; "  << std::endl
              << "std::optional has no map() or filter() like Java" << std::endl;
}

void direct_value_access()
{
    std::cout << "direct_value_access" << std::endl
              << "-------------------" << std::endl;

    // Given
    auto hyunday_opt = std::make_optional(loud_car);

    // When

    // (1) value() is not the only way to get to the value.
    auto manufacturer1 = hyunday_opt.value().manufacturer;
    // (2) you can also dereference it with *
    auto manufacturer2 = (*hyunday_opt).manufacturer;
    // (3) you can also dereference it with ->
    auto manufacturer3 = hyunday_opt->manufacturer;

    // Then
    if (manufacturer1 == manufacturer2 
            && manufacturer2 == manufacturer3
            && manufacturer3 == "Hyundai") {
        std::cout << "C++ offers some nicer ways to directly access the object" 
                  << std::endl;
    }
}

int main() 
{
    optional_of_null();
    std::cout << std::endl << std::endl;

    optional_of_null_usage();
    std::cout << std::endl << std::endl;

    optional_of_value_usage();
    std::cout << std::endl << std::endl;
    
    optional_of_value_other_usage();
    std::cout << std::endl << std::endl;
    
    bouncer_patterns();
    std::cout << std::endl << std::endl;
    
    fake_repository_return_value();
    std::cout << std::endl << std::endl;
    
    direct_value_access();
    std::cout << std::endl << std::endl;
    
    return 0;
}

File: java/src/test/java/com/thecodeslinger/OptionalTest.java
package com.thecodeslinger;

import lombok.AllArgsConstructor;
import lombok.Data;
import org.junit.jupiter.api.Test;

import java.util.NoSuchElementException;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.catchThrowable;

/**
 * The test cases use the same names as the test methods in the C++ code, only styled
 * to match Java conventions.
 *
 * The numbers in parenthesis, e.g. (1) or (2) match the C++ code samples
 * in the respective test cases. Missing numbers mean missing features.
 *
 * This pattern is used where applicable.
 */
class OptionalTest {

    private static final Car LOUD_CAR = new Car("Hyundai", "i30 N");
    private static final Car REASONABLE_CAR = new Car("Volkswagen", "Golf");
    private static final Car HUGE_CAR = new Car("Ford", "Raptor");

    /**
     * A custom class to demonstrate the use of {@link Optional}
     */
    @Data
    @AllArgsConstructor
    static class Car {
        private String manufacturer;
        private String model;
    }

    /**
     * Another custom class to demonstrate the use of {@link Optional}
     */
    @Data
    @AllArgsConstructor
    static class Truck {
        private String manufacturer;
        private String model;
        private String size;
    }

    @Test
    void optionalOfNull() {
        // When

        // Optional.of expects non-null.
        var nullPtrException = catchThrowable(() -> Optional.of(null));
        // Optional.ofNullable works with null.
        var nullOpt = Optional.ofNullable(null);
        // Equivalent to above; usually used as return value.
        var emptyOpt = Optional.empty();

        // Then
        assertThat(nullPtrException).isInstanceOf(NullPointerException.class);
        assertThat(nullOpt).isEmpty();
        assertThat(emptyOpt).isEmpty();
    }

    @Test
    void optionalOfNullUsage() {
        // Given
        var nullOpt = Optional.<Car>empty();

        // When

        // (1) Get the optional or another optional if null.
        var nullOr = nullOpt.or(OptionalTest::produceTruck);
        // (2) Get the actual value of the optional or another value if null.
        var nullOrElse = nullOpt.orElse(REASONABLE_CAR);
        // (3) Get the actual value of the optional or throw an exception if null. "Bouncer pattern".
        var nullOrElseThrow = catchThrowable(() -> nullOpt.orElseThrow(IllegalArgumentException::new));
        // (4) Perform an action if the optional is not empty.
        var nullIfPresent = catchThrowable(() -> nullOpt.ifPresent(OptionalTest::haveFun));
        // (5) get() on empty optional throws.
        var nullGet = catchThrowable(() -> nullOpt.get());

        // Then

        // (1)
        assertThat(nullOr)
                .isInstanceOf(Optional.class)
                .isNotEmpty();
        assertThat(nullOr.get()).isEqualTo(HUGE_CAR);
        // (2)
        assertThat(nullOrElse).isEqualTo(REASONABLE_CAR);
        // (3)
        assertThat(nullOrElseThrow).isInstanceOf(IllegalArgumentException.class);
        // (4) Null because ifPresent does not call haveFun if optional is empty.
        assertThat(nullIfPresent).isNull();
        // (5)
        assertThat(nullGet).isInstanceOf(NoSuchElementException.class);
    }

    @Test
    void optionalOfValueUsage() {
        // Given
        var hyundaiOpt = Optional.of(LOUD_CAR);

        // Expect
        assertThat(hyundaiOpt).isNotEmpty();

        // When

        // (1) This time the LOUD_CAR is returned.
        var hyundaiOr = hyundaiOpt.or(OptionalTest::produceTruck);
        // (2) Again, LOUD_CAR.
        var hyundaiOrElse = hyundaiOpt.orElse(REASONABLE_CAR);
        // (3) No exception.
        var hyundaiOrElseThrow = hyundaiOpt.orElseThrow(IllegalArgumentException::new);
        // (4) haveFun() is called, hence an exception is thrown.
        var hyundaiIfPresent = catchThrowable(() -> hyundaiOpt.ifPresent(OptionalTest::haveFun));

        // Then

        // (1)
        assertThat(hyundaiOr).isEqualTo(hyundaiOpt);
        // (2)
        assertThat(hyundaiOrElse).isEqualTo(LOUD_CAR);
        // (3)
        assertThat(hyundaiOrElseThrow).isEqualTo(LOUD_CAR);
        // (4)
        assertThat(hyundaiIfPresent)
                .isInstanceOf(IllegalStateException.class)
                .hasMessage("Hyundai i30 N");
    }

    @Test
    void optionalOfValueOtherUsage() {
        // Given
        var nullOpt = Optional.<Car>empty();
        var hyundaiOpt = Optional.of(LOUD_CAR);

        // When

        // (1) Convert to another type. This example only extract a property.
        var manufacturer = hyundaiOpt.map(Car::getManufacturer);
        // (2) The same, only with an empty nullable.
        var nullMap = nullOpt.map(Car::getManufacturer);
        // (3) Only let certain values pass.
        var filtered = hyundaiOpt.filter(car -> car.getModel().equalsIgnoreCase("i30 N"));


        // Then

        // (1)
        assertThat(manufacturer).isEqualTo(Optional.of("Hyundai"));
        // (2)
        assertThat(nullMap).isEmpty();
        // (3)
        assertThat(filtered).isEqualTo(hyundaiOpt);
    }

    @Test
    void bouncerPatterns() {
        // Given
        var hyundaiOpt = Optional.of(LOUD_CAR);


        // (1) Use isPresent() or isEmpty() condition. Most obvious and flexible approach.
        if (hyundaiOpt.isEmpty()) {
            // throw ...
            assertThat(false).isTrue(); // Not called.
        }
        var value1 = hyundaiOpt.get();
        assertThat(value1).isEqualTo(LOUD_CAR);

        // (2) Use orElseThrow(). No condition and immediately returns value.
        var value2 = hyundaiOpt.orElseThrow(NullPointerException::new);
        assertThat(value2).isEqualTo(LOUD_CAR);
    }

    @Test
    void fakeRepositoryReturnValue() {
        // Given
        var userSearchString = "raptor";

        // When
        var truck = repository_findByModel(userSearchString)
                .map(OptionalTest::fromCar)
                .orElseThrow(IllegalArgumentException::new);

        // Then
        assertThat(truck).isInstanceOf(Truck.class);
    }

    @Test
    void directValueAccess() {
        // Given
        var hyundaiOpt = Optional.of(LOUD_CAR);

        // When

        // (1) get() is the only way to get to the value.
        var manufacturer = hyundaiOpt.get().getManufacturer();

        // Then

        // (1)
        assertThat(manufacturer).isEqualTo("Hyunday");

    }

    /*********** Helpers ***********/

    static Optional<Car> produceTruck() {
        return Optional.of(new Car("Ford", "Raptor"));
    }

    static void haveFun(final Car car) {
        if (null == car) {
            throw new IllegalArgumentException("This ain't not fun");
        }

        // Used to be able to verify the paths taken in this method.
        throw new IllegalStateException(car.getManufacturer() + " " + car.getModel());
    }

    static Optional<Car> repository_findByModel(final String model) {
        if (model.equalsIgnoreCase("Raptor")) {
            return Optional.of(HUGE_CAR);
        }
        return Optional.empty();
    }

    static Truck fromCar(final Car car) {
        return new Truck(car.getManufacturer(), car.getModel(), "HUGE");
    }
}
